### PR TITLE
feat(ingest): add output schema inference for sql parser

### PR DESCRIPTION
**Pull request by hsheth2**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/utilities/sqlglot_lineage.py [hunk 1]
```diff
@@ -541,6 +565,15 @@ def _schema_aware_fuzzy_column_resolve(
...
     logger.debug("Qualified sql %s", statement.sql(pretty=True, dialect=dialect))
 
+    # Try to figure out the types of the output columns.
+    try:
+        statement = sqlglot.optimizer.annotate_types.annotate_types(
+            statement, schema=sqlglot_db_schema
+        )
+    except sqlglot.errors.OptimizeError as e:
+        # This is not a fatal error, so we can continue.
+        logger.debug("sqlglot failed to annotate types: %s", e)
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/utilities/sqlglot_lineage.py:585]: Could this be slow? I think it'd be nice to only do this if a config option is specified

File: metadata-ingestion/tests/integration/powerbi/test_m_parser.py [hunk 2]
```diff
@@ -742,75 +742,25 @@ def test_sqlglot_parser():
...
-            logic=None,
-        ),
-        ColumnLineageInfo(
-            downstream=DownstreamColumnRef(table=None, column="software_quota"),
-            upstreams=[],
-            logic=None,
-        ),
+    # TODO: None of these columns have upstreams?
+    # That doesn't seem right - we probably need to add fake schemas for the two tables above.
+    cols = [
```
Comment (asikowitz) [inline metadata-ingestion/tests/integration/powerbi/test_m_parser.py:747]: Nice

**Review by asikowitz — APPROVED** [review]
I'm not following how some of these methods get called. Just want to make sure there's no performance concerns here, as I know sql parsing can be kinda slow as is

File: metadata-ingestion/src/datahub/utilities/sqlglot_lineage.py [hunk 3]
```diff
@@ -90,8 +102,18 @@ def get_query_type_of_sql(expression: sqlglot.exp.Expression) -> QueryType:
     return QueryType.UNKNOWN
 
 
+class _ParserBaseModel(
+    BaseModel,
+    arbitrary_types_allowed=True,
+    json_encoders={
+        SchemaFieldDataTypeClass: lambda v: v.to_obj(),
+    },
```
[the remark this answers is not in the document]
Comment (hsheth2) [reply metadata-ingestion/src/datahub/utilities/sqlglot_lineage.py:109]: I'm honestly not too happy with this setup, but it's fine for now

File: metadata-ingestion/src/datahub/utilities/sqlglot_lineage.py (hunk 1 above)
Comment (hsheth2) [reply metadata-ingestion/src/datahub/utilities/sqlglot_lineage.py:585]: this step should be pretty fast

**Comment by hsheth2** [comment]
Merging through flaky smoke tests.

**Pull request merged**